variable_set.go

```diff
@@ -0,0 +1,255 @@
+	if err != nil {
+		return nil, err
+	}
+
+	vs := &VariableSet{}
+	err = s.client.do(ctx, req, vs)
+	if err != nil {
+		return nil, err
+	}
+
+	return vs, err
+}
+
+// VariableSetUpdateOptions represents the options for updating a variable set.
+type VariableSetUpdateOptions struct {
+	// Type is a public field utilized by JSON:API to
+	// set the resource type via the field tag.
+	// It is not a user-defined value and does not need to be set.
+	// https://jsonapi.org/format/#crud-creating
+	Type string `jsonapi:"primary,vars"`
+
+	// The name of the variable set.
+	// Affects variable precedence when there are conflicts between Variable Sets
+	// https://www.terraform.io/cloud-docs/api-docs/variable-sets#apply-variable-set-to-workspaces
+	Name *string `jsonapi:"attr,name"`
+
+	// A description to provide context for the variable set.
+	Description *string `jsonapi:"attr,description"`
+
+	// If true the variable set is considered in all runs in the organization.
+	Global *bool `jsonapi:"attr,global"`
+}
+
+func (s *variableSets) Update(ctx context.Context, variableSetID string, options VariableSetUpdateOptions) (*VariableSet, error) {
+	if !validStringID(&variableSetID) {
+		return nil, errors.New("invalid value for variable set ID")
+	}
+
+	u := fmt.Sprintf("varsets/%s?include=workspaces", url.QueryEscape(variableSetID))
+	req, err := s.client.newRequest("PATCH", u, &options)
+	if err != nil {
+		return nil, err
+	}
+
+	v := &VariableSet{}
+	err = s.client.do(ctx, req, v)
+	if err != nil {
+		return nil, err
+	}
+
+	return v, nil
+}
+
+// Delete a variable set by its ID.
+func (s *variableSets) Delete(ctx context.Context, variableSetID string) error {
+	if !validStringID(&variableSetID) {
+		return errors.New("invalid value for variable set ID")
+	}
+
+	u := fmt.Sprintf("varsets/%s", url.QueryEscape(variableSetID))
+	req, err := s.client.newRequest("DELETE", u, nil)
+	if err != nil {
+		return err
+	}
+
+	return s.client.do(ctx, req, nil)
+}
+
+// VariableSetAssignOptions represents a subset of update options specifically for assigning variable sets to workspaces
+type VariableSetAssignOptions struct {
+	// Type is a public field utilized by JSON:API to
+	// set the resource type via the field tag.
+	// It is not a user-defined value and does not need to be set.
+	// https://jsonapi.org/format/#crud-creating
+	Type string `jsonapi:"primary,vars"`
+
+	// Used to set the variable set from Global to not Global if necessary
+	Global *bool `jsonapi:"attr,global"`
+
+	// The workspaces to be assigned to. An empty set means remove all assignments
+	Workspaces []*Workspace `jsonapi:"relation,workspaces"`
+}
+
+// Use Update to assign a variable set to workspaces
+func (s *variableSets) Assign(ctx context.Context, variableSetID string, options VariableSetAssignOptions) (*VariableSet, error) {
+	if options.Workspaces == nil {
+		return nil, errors.New("No workspaces list provided")
+	}
+
+	options.Global = Bool(false)
```

Comment:
Maybe this should not be an attribute on the assign options since you set it statically? Is it necessary to always pass false to this endpoint?

Reply:
For this case, Yes. The implication that you want to assign specific workspaces means you no longer want the variable set to be global.

If not set to false, and the variable set is global, workspace assignment will fail.